metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala

```diff
@@ -118,21 +118,31 @@ object FileWatcher {
       watchFilter: Path => Boolean
   ): () => Unit = {
     if (scala.util.Properties.isMac) {
-      // Due to a hard limit on the number of FSEvents streams that can be opened on macOS,
-      // only the root workspace directory is registered for a recursive watch.
+      // Due to a hard limit on the number of FSEvents streams that can be
+      // opened on macOS, only up to 32 longest common prefixes of the files to
+      // watch are registered for a recursive watch.
       // However, the events are then filtered to receive only relevant events
-      // and also to hash only revelevant files when watching for changes
+      // and also to hash only relevant files when watching for changes
 
       val trie = PathTrie(
         filesToWatch.sourceFiles ++ filesToWatch.sourceDirectories ++ filesToWatch.semanticdDirectories
       )
       val isWatched = trie.containsPrefixOf _
 
+      // Select up to `maxRoots` longest prefixes of all files in the trie for
+      // watching. Watching the root of the workspace may have bad performance
+      // implications if it contains many other projects that we don't need to
+      // watch (eg. in a monorepo)
+      val watchRoots = trie.longestPrefixes(workspace.getRoot(), maxRoots = 32)
```

Comment:
The `32` here is completely arbitrary and worked well for us in the projects where I tried this change. What's the policy around adding configuration knobs? Are system properties the preferred option?